File: src/main/java/net/gegy1000/terrarium/server/map/source/GeocodingSource.java
package net.gegy1000.terrarium.server.map.source;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import net.gegy1000.terrarium.Terrarium;
import net.gegy1000.terrarium.server.util.Coordinate;
import net.gegy1000.terrarium.server.world.EarthGenerationSettings;

import javax.net.ssl.HttpsURLConnection;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.net.URLEncoder;

public class GeocodingSource implements DataSource {
    private static final JsonParser JSON_PARSER = new JsonParser();

    private final EarthGenerationSettings settings;

    public GeocodingSource(EarthGenerationSettings settings) {
        this.settings = settings;
    }

    public Coordinate get(String place) throws IOException {
        String request = "https://maps.googleapis.com/maps/api/geocode/json?address=" + URLEncoder.encode(place, "UTF-8");

        URL url = new URL(request);
        HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
        connection.setRequestMethod("GET");
        connection.setRequestProperty("User-Agent", Terrarium.MODID);

        try (InputStreamReader input = new InputStreamReader(connection.getInputStream())) {
            JsonObject root = (JsonObject) JSON_PARSER.parse(input);
            if (root.has("results")) {
                JsonArray results = root.getAsJsonArray("results");
                for (JsonElement element : results) {
                    JsonObject result = element.getAsJsonObject();
                    if (result.has("geometry")) {
                        JsonObject geometry = result.getAsJsonObject("geometry");
                        JsonObject location = geometry.getAsJsonObject("location");

                        return Coordinate.fromLatLng(this.settings, location.get("lat").getAsDouble(), location.get("lng").getAsDouble());
                    }
                }
            }
        }

        return null;
    }

    @Override
    public EarthGenerationSettings getSettings() {
        return this.settings;
    }
}

File: src/main/kotlin/net/gegy1000/terrarium/server/map/source/GeocodingSource.kt
package net.gegy1000.terrarium.server.map.source

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import net.gegy1000.terrarium.Terrarium
import net.gegy1000.terrarium.server.util.Coordinate
import net.gegy1000.terrarium.server.world.EarthGenerationSettings
import java.io.IOException
import java.io.InputStreamReader
import java.net.URL
import java.net.URLEncoder
import javax.net.ssl.HttpsURLConnection

class GeocodingSource(override val settings: EarthGenerationSettings) : DataSource {
    private val JSON_PARSER = JsonParser()

    fun get(place: String): Coordinate? {
        val request = "https://maps.googleapis.com/maps/api/geocode/json?address=${URLEncoder.encode(place, "UTF-8")}"

        val url = URL(request)
        val connection = url.openConnection() as HttpsURLConnection
        connection.requestMethod = "GET"
        connection.setRequestProperty("User-Agent", Terrarium.MODID)

        val input = connection.inputStream

        try {
            val root = JSON_PARSER.parse(InputStreamReader(input)) as JsonObject

            if (root.has("results")) {
                val results = root["results"].asJsonArray

                results.map { it.asJsonObject }.filter { it.has("geometry") }.forEach { result ->
                    val geometry = result["geometry"].asJsonObject
                    val location = geometry["location"].asJsonObject

                    return Coordinate.fromLatLng(settings, location["lat"].asDouble, location["lng"].asDouble)
                }
            }
        } catch (e: IOException) {
            Terrarium.LOGGER.error("Failed to retrieve geocode for $place", e)
        } finally {
            input.close()
        }

        return null
    }
}
